lib/ui/platform_dispatcher.dart

```diff
@@ -2297,3 +2342,155 @@ enum DartPerformanceMode {
+    assert(minScaleFactor.isFinite);
+    assert(minScaleFactor >= 0);
+
+    return minScaleFactor == maxScaleFactor
+         ? TextScaler.linear(minScaleFactor)
+         : _ClampedTextScaler(this, minScaleFactor, maxScaleFactor);
+  }
+}
+
+final class _LinearTextScaler implements TextScaler {
+  const _LinearTextScaler(this.textScaleFactor) : assert(textScaleFactor > 0);
+
+  final double textScaleFactor;
+
+  @override
+  double scale(double fontSize) {
+    assert(fontSize >= 0);
+    assert(fontSize.isFinite);
+    return fontSize * textScaleFactor;
+  }
+
+  @override
+  TextScaler clamp(double minScale, double maxScale) {
+    final double newScaleFactor = clampDouble(textScaleFactor, minScale, maxScale);
+    return newScaleFactor == textScaleFactor ? this : _LinearTextScaler(newScaleFactor);
+  }
+
+  @override
+  bool operator ==(Object other) {
+    if (identical(this, other)) {
+      return true;
+    }
+    return other is _LinearTextScaler && other.textScaleFactor == textScaleFactor;
+  }
+
+  @override
+  int get hashCode => textScaleFactor.hashCode;
+}
+
+final class _ClampedTextScaler implements TextScaler {
+  const _ClampedTextScaler(this.scaler, this.minScale, this.maxScale) : assert(maxScale > minScale);
+  final TextScaler scaler;
+  final double minScale;
+  final double maxScale;
+
+  @override
+  double scale(double fontSize) {
+    assert(fontSize >= 0);
+    assert(fontSize.isFinite);
+    return minScale == maxScale
+      ? minScale * fontSize
+      : clampDouble(scaler.scale(fontSize), minScale * fontSize, maxScale * fontSize);
+  }
+
+  @override
+  TextScaler clamp(double minScale, double maxScale) {
+    return minScale == maxScale
+      ? _LinearTextScaler(minScale)
+      : _ClampedTextScaler(scaler, math.max(minScale, this.minScale), math.min(maxScale, this.maxScale));
+  }
+
+  @override
+  bool operator ==(Object other) {
+    if (identical(this, other)) {
+      return true;
+    }
+    return other is _ClampedTextScaler
+        && minScale == other.minScale
+        && maxScale == other.maxScale
+        && (minScale == maxScale || scaler == other.scaler);
+  }
+
+  @override
+  int get hashCode => minScale == maxScale ? minScale.hashCode : Object.hash(scaler, minScale, maxScale);
+}
+
+final class _SystemTextScaler extends TextScaler {
+  const _SystemTextScaler(this.textScaleFactor);
+
+  final double textScaleFactor;
+
+  @override
+  double scale(double fontSize) {
+    assert(fontSize >= 0);
+    assert(fontSize.isFinite);
+    switch (PlatformDispatcher._getScaledFontSize(fontSize)) {
+      case >= 0 && final double x when fontSize >= 0:
+        return x;
+      case == -1 when fontSize >= 0:
+        throw StateError('GetScaledFontSize is not implemented on the platform. This should not be reached.');
+      // Fallback to linear scaling when there's an error. Relevant error messages
+      // should already be printed to stderr.
+      case final double errorCode:
```

Comment:
Why does this case exist?

Reply:
currently the android implementation returns -2 if the delegate is not set, for example.